platform/commonUI/edit/src/services/TransactionService.js

```diff
@@ -0,0 +1,118 @@
+/*****************************************************************************
+ * Open MCT Web, Copyright (c) 2014-2015, United States Government
+ * as represented by the Administrator of the National Aeronautics and Space
+ * Administration. All rights reserved.
+ *
+ * Open MCT Web is licensed under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * Open MCT Web includes source code licensed under additional open source
+ * licenses. See the Open Source Licenses file (LICENSES.md) included with
+ * this source code distribution or the Licensing information page available
+ * at runtime from the About dialog for additional information.
+ *****************************************************************************/
+/*global define*/
+define(
+    [],
+    function() {
+        /**
+         * Implements an application-wide transaction state. Once a
+         * transaction is started, calls to PersistenceCapability.persist()
+         * will be deferred until a subsequent call to
+         * TransactionService.commit() is made.
+         *
+         * @param $q
+         * @constructor
+         */
+        function TransactionService($q, dirtyModelCache) {
+            this.$q = $q;
+            this.transaction = false;
+            this.committing = false;
+            this.cache = dirtyModelCache;
+        }
+
+        TransactionService.prototype.startTransaction = function () {
+            if (this.transaction)
+                console.error("Transaction already in progress")
+            this.transaction = true;
+        };
+
+        TransactionService.prototype.isActive = function () {
+            return this.transaction;
+        };
+
+        TransactionService.prototype.isCommitting = function () {
+            return this.committing;
+        };
+
+        /**
+         * All persist calls deferred since the beginning of the transaction
+         * will be committed. Any failures will be reported via a promise
+         * rejection.
+         * @returns {*}
+         */
+        TransactionService.prototype.commit = function () {
+            var self = this;
+                cache = this.cache.get();
+
+            this.committing = true;
+
+            function keyToObject(key) {
+                return cache[key];
+            }
+
+            function objectToPromise(object) {
+                return object.getCapability('persistence').persist();
```

Comment:
If I'm understanding the flow right, the decorated `persist` gets called here, and doesn't just stick the object back in the dirty cache because `isCommitting()` is true at that point. This strikes me as a bit of a delicate balance between the three components (transaction service, persistence capability decorator, and dirty model queue)

What if, instead of getting objects from the queue, transactionService just let you push functions into it:

`transactionService.addToTransaction(onCommit, onCancel)`

When `commit` is called, it invokes all the `onCommits`; when `cancel` is called, it invokes all the `onCancels`. In either case, when it's done it clears the `active` flag and empties its queues. e.g.:

```
TransactionService.prototype.commit = function () {
    while (this.onCommits.length > 0) { // ...using a while in case some onCommit adds to transaction
        try { // ...also don't want to fail mid-loop...
            this.onCommits.unshift()();
        } catch (e) {
            this.$log.error("Error committing transaction.");
        }
    }
    this.active = false;
    this.onCommits = [];
    this.onCancels = [];
};
```

This would reduce TransactionService's dependencies greatly, making it much more testable, and less sensitive to any external changes.